contrib/mesos/pkg/scheduler/service/util.go

```diff
@@ -86,3 +94,131 @@ func hashExecutorInfo(info *mesos.ExecutorInfo) uint64 {
+	go func() {
+		defer close(out)
+		for entry := range in {
+			if entry.Dir {
+				continue
+			}
+			if entry.Err != nil {
+				log.Errorf("error fetching static pod file %q: %v", entry.Path, entry.Err)
+				continue
+			}
+			keep := false
+			defaultFn := func(pod *api.Pod) (err error) {
+				keep, err = filter(pod)
+				return
+			}
+			parsed, pod, err := tryDecodeSinglePod(entry.Data, defaultFn)
+			if !parsed {
+				if err != nil {
+					entry.Err = err
+					log.Errorf("error parsing static pod file %q: %v", entry.Path, err)
+				}
+				continue
+			}
+			if err != nil {
+				entry.Err = err
+				log.Errorf("error validating static pod file %q: %v", entry.Path, err)
+				continue
+			}
+			if keep {
+				// marshal back into JSON form to pick up any changes made to
+				// the pod specification
+				updated, err := api.Scheme.EncodeToVersion(pod, "v1")
+				if err != nil {
+					entry.Err = err
+					log.Errorf("error serializing pod json for %v/%v: %v", pod.Namespace, pod.Name, err)
+					continue
+				}
+				entry.Data = updated
+				out <- entry
+			}
+		}
+	}()
+	return out
+}
+
+func makeStaticPodValidator(limitCPU resource.CPUShares, limitMem resource.MegaBytes, accumCPU, accumMem *float64, minimalResources bool) filterFunc {
+	return filterFunc(func(pod *api.Pod) (bool, error) {
+		// TODO(sttts): allow unlimited static pods as well and patch in the default resource limits
+		unlimitedCPU := resource.LimitPodCPU(pod, limitCPU)
+		if unlimitedCPU {
+			return false, fmt.Errorf("found static pod without limit on cpu resources")
+		}
+
+		unlimitedMem := resource.LimitPodMem(pod, limitMem)
+		if unlimitedMem {
+			return false, fmt.Errorf("found static pod without limit on memory resources")
+		}
+
+		cpu := resource.PodCPULimit(pod)
+		mem := resource.PodMemLimit(pod)
+		if minimalResources {
+			// see SchedulerServer.AccountForPodResources
+			cpu = podtask.MinimalCpus
+			mem = podtask.MinimalMem
+		}
+
+		log.V(2).Infof("reserving %.2f cpu shares and %.2f MB of memory to static pod %s/%s", cpu, mem, pod.Namespace, pod.Name)
+
+		*accumCPU += float64(cpu)
+		*accumMem += float64(mem)
+		return true, nil
+	})
+}
+
+func makeStaticPodProcurement(entries []*archive.Entry) podtask.Procurement {
+	if len(entries) == 0 {
+		return nil
+	}
+
+	// entries is a pre-validated list of static pod specs that serve as templates for the static pod builder func
+	return podtask.StaticPodProcurement(func(assignedSlave string) ([]byte, error) {
+		applyBindingsTo := filterFunc(func(pod *api.Pod) (bool, error) {
+			if pod.Annotations == nil {
+				pod.Annotations = make(map[string]string)
+			}
+			// add binding annotation so that static pods aren't picked up by the main scheduling loop.
+			// static pods are always completely managed by the kubelet.
+			pod.Annotations[meta.BindingHostKey] = assignedSlave
+			return true, nil
+		})
+		return archive.Create(applyBindingsTo.podEntries(archive.Playback(entries)))
```

Comment:
that's a very good point

On Fri, Sep 18, 2015 at 12:03 PM, Dr. Stefan Schimanski <
notifications@github.com> wrote:

> In contrib/mesos/pkg/scheduler/service/util.go
> <https://github.com/kubernetes/kubernetes/pull/13943#discussion_r39871470>
> :
>
> > +	if len(entries) == 0 {
> > +		return nil
> > +	}
> > +
> > +	// entries is a pre-validated list of static pod specs that serve as templates for the static pod builder func
> > +	return podtask.StaticPodProcurement(func(assignedSlave string) ([]byte, error) {
> > +		applyBindingsTo := filterFunc(func(pod *api.Pod) (bool, error) {
> > +			if pod.Annotations == nil {
> > +				pod.Annotations = make(map[string]string)
> > +			}
> > +			// add binding annotation so that static pods aren't picked up by the main scheduling loop.
> > +			// static pods are always completely managed by the kubelet.
> > +			pod.Annotations[meta.BindingHostKey] = assignedSlave
> > +			return true, nil
> > +		})
> > +		return archive.Create(applyBindingsTo.podEntries(archive.Playback(entries)))
>
> If we use the pod list, we can easily do the binding modification in the
> executor instead of pulling this into the procurement.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13943/files#r39871470>.
>